src/rusoto/auth.rs

```diff
@@ -0,0 +1,92 @@
+use super::AwsCredentialsProvider;
+use rusoto_core::Region;
+use serde::{Deserialize, Serialize};
+
+/// Configuration for configuring authentication strategy for AWS.
+#[derive(Serialize, Deserialize, Clone, Debug, Derivative)]
+#[serde(untagged)]
+#[derivative(Default)]
+pub enum AWSAuthentication {
+    Role {
+        assume_role: String,
+    },
+    // Default variant is used instead of Option<AWSAuthentication> since even for
+    // None we need to build `AwsCredentialsProvider`.
+    //
+    // {} is required to work around a bug in serde. https://github.com/serde-rs/serde/issues/1374
+    #[derivative(Default)]
+    Default {},
+}
+
+impl AWSAuthentication {
+    pub fn build(
+        &self,
+        region: &Region,
+        old_assume_role: Option<String>,
+    ) -> crate::Result<AwsCredentialsProvider> {
+        if old_assume_role.is_some() {
+            warn!("Option `assume_role` has been renamed to `auth.assume_role`. Please use that one instead.");
+        }
+        match self {
+            Self::Role { assume_role } => {
+                if old_assume_role.is_some() {
+                    warn!("Ignoring option `assume_role` and using option `auth.assume_role` instead.");
+                }
+                AwsCredentialsProvider::new(region, Some(assume_role.clone()))
+            }
+            Self::Default {} => AwsCredentialsProvider::new(region, old_assume_role),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[derive(Serialize, Deserialize, Clone, Debug)]
+    struct ComponentConfig {
+        assume_role: Option<String>,
+        #[serde(default)]
+        auth: AWSAuthentication,
+    }
+
+    #[test]
+    fn parsing_default() {
+        toml::from_str::<ComponentConfig>(
+            r#"
+        "#,
+        )
+        .unwrap();
+    }
+
+    #[test]
+    fn parsing_old_assume_role() {
+        toml::from_str::<ComponentConfig>(
+            r#"
+            assume_role = "root"
+        "#,
+        )
+        .unwrap();
+    }
+
+    #[test]
+    fn parsing_assume_role() {
+        toml::from_str::<ComponentConfig>(
+            r#"
+            auth.assume_role = "root"
+        "#,
+        )
+        .unwrap();
+    }
+
+    #[test]
+    fn parsing_both_assume_role() {
+        toml::from_str::<ComponentConfig>(
```

Comment:
Should we assert that it chooses `auth.assume_role` here? We could add assertions to the others as well. I see we'd really just be asserting that serde does what it should but we've been surprised by it before. This one is actually logic that we have though.

Reply:
One other test will fail in that case, but yea, better be sure and also test that here.